contracts/TaskMgt.sol

```diff
@@ -0,0 +1,263 @@
+     * @param result The computing result content including zk proof.
+     * @return True if reporting is successful.
+     */
+    function reportResult(bytes32 taskId, bytes calldata result) external returns (bool) {
+        // TODO
+        bytes32 workerId = keccak256(abi.encode(msg.sender));
+        Task storage task = _allTasks[taskId];
+        require(task.taskId == taskId, "the task does not exist");
+
+        require(task.status == TaskStatus.PENDING, "the task status is not PENDING");
+        ComputingInfo storage computingInfo = task.computingInfo;
+
+        uint256 waitingIndex = find(workerId, computingInfo.waitingList);
+        require(waitingIndex != type(uint256).max, "task id not in waiting list");
+
+        uint256 workerIndex = find(workerId, computingInfo.workerIds);
+        computingInfo.results[workerIndex] = result;
+
+        uint256 waitingListLength = computingInfo.waitingList.length;
+        computingInfo.waitingList[waitingIndex] = computingInfo.waitingList[waitingListLength - 1];
+        computingInfo.waitingList.pop();
+        waitingListLength--;
+
+        bytes32[] storage taskIds = _taskIdForWorker[workerId];
+        uint256 taskIndex = find(taskId, taskIds);
+        taskIds[taskIndex] = taskIds[taskIds.length - 1];
+        taskIds.pop();
+
+        if (waitingListLength == 0) {
+            uint256 pendingIndex = find(taskId, _pendingTaskIds);
+            _pendingTaskIds[pendingIndex] = _pendingTaskIds[_pendingTaskIds.length - 1];
+            _pendingTaskIds.pop();
+
+            _completedTaskIds.push(taskId);
+
+            task.status = TaskStatus.COMPLETED;
+            emit TaskCompleted(task.taskId);
+        }
+
+        return true;
+    }
+
+    /**
+     * @notice Get the tasks that need to be run by Workers.
+     * @return Returns an array of tasks that the workers will run.
+     */
+    function getPendingTasks() external view returns (Task[] memory) {
+        uint256 taskCount = _pendingTaskIds.length;
+
+        Task[] memory tasks = new Task[](taskCount);
+        for (uint256 i = 0; i < taskCount; i++) {
+            tasks[i] = _allTasks[_pendingTaskIds[i]];
+        }
+
+        return tasks;
+    }
+
+    /**
+     * @notice Get the tasks that a Worker needs to run.
+     * @param workerId The Worker id.
+     * @return Returns an array of tasks that the worker will run.
+     */
+    function getPendingTasksByWorkerId(bytes32 workerId) external view returns (Task[] memory) {
+        bytes32[] storage taskIds = _taskIdForWorker[workerId];
+        uint256 taskCount = taskIds.length;
+
+        Task[] memory tasks = new Task[](taskCount);
+        for (uint i = 0; i < taskCount; i++) {
+            tasks[i] = _allTasks[taskIds[i]];
+        }
+
+        return tasks;
+    }
+
+    /**
+     * @notice Get completed tasks.
+     * @return Returns an array of completed tasks.
+     */
+    function getCompletedTasks() external view returns (Task[] memory) {
```

Comment:
Because the completed tasks may be very big, so this function can delete and can use subgraph to query all tasks. The _completedTaskIds data filed can be delete.